### DISCUSSION TITLE
Support / Implementation for a new Data Type: Pair (and others)

### DISCUSSION THREAD

**Discussion by bernarma** [josdejong/mathjs#2163] [Q&A]
**Background**
I am considering adding a new type called "Pair" (among other changes) to represent an x,y coordinate that I would then like to use in a graph/chart to create one or more series.

var1 = (1,2) // x,y coordinates
var2 = (,3) // y coordinates specified - represents horizontal line
var3 = (3,) // x coordinates specified - represents vertical line

These points may not necessarily be a numeric for the x or the y as I would like to introduce a date and time data type and potentially others down the track. Also, I would then proceed and add support to the relevant functions to support the operations on this data type.

Hopefully I have set the background of what I am looking at achieving.

**Question**
Based on the background of what I am trying to achieve, I would like to understand the following:

- is something the project would like to see as a feature (or should it be kept separate)
- are the any issues / complications that you see in the implementation of this feature?

I guess I am looking for thoughts on this before I go off and implement it and potentially see if there are other thoughts that could achieve the above without writing a heap of code.

**Comment by josdejong** [comment]
Hi Mark, 

A `Pair` can be useful but is relatively limited. I think it's similar to an Array or Matrix with just two values?

If you I want to hold more information together (x, y, date, ...), using an Object may be a good solution (that makes it explicit what value has what meaning, instead of relying on the position of a value in a Pair, Tuple, or Array.

Would something like this work for you?

```
var1 = { x: 1, y: 2 }
var2 = { x: null, y: 3 }
var2 = { x: 3, y: null }
```

**Reply by bernarma** [reply]
Thanks for responding. I gave it some more thought and have played with a few ideas and agree that the pair may be quite limited albeit a more simplistic format for user entry. You are correct in that it is just a two value matrix / array. I do need to incorporate units into the equation as well which I haven't yet played with but will do so over the next few days. I have found this library so far really enjoyable to work with. I am currently working on a proof of concept at the moment. I will need to see how the performance goes in the short-medium term as I will be working with quite large datasets. I may end up looking into the gpu.js library if performance is an issue (I have seen a mention in the questions from others) - but no promises :). Either way, performance is secondary at the moment and the functionality / suitability is the key to start with.

**Reply by josdejong** [reply]
👍  sounds good. For larger data sets you can probably simply use a Matrix/Array, where for example the columns represent `x` and `y`, and the rows represent different items.

**Reply by bernarma** [reply]
I am experimenting with a couple of ideas, namely, using the Matrix/Array and the Object.

I have defined a "Custom Type" called PairType with an x and y property. I have also provided custom functions where that type is recognised. What I have noticed though is I am unable to parse an expression whereby the custom type is recognised.

expression := { x: 10, y:100, isPairType:true } does not equal new math.PairType(10, 100). Is there a way for the expression parser to recognise newly defined types or can you suggest another way of achieving the same result?

**Reply by josdejong** [reply]
Yes, you can implement new data types for mathjs, see this example: https://mathjs.org/examples/advanced/custom_datatype.js.html

**Reply by bernarma** [reply]
Sorry @josdejong , I wasn't clear.

I have defined a Custom Type and have been able to successfully use that type in calculations as described in the docs.

My issue is, when I attempt to use > math.parse("sum([{x: 1, y:150, isPairType: true}, {x: 2, y:100, isPairType: true}])"), it is treating that object as an object, it is not parsing it into a custom type called PairType resulting in the sum function. to fail.

Is there any way this can be achieved without touching the parser code in the library?

I know the example doesn't particularly make sense but the idea is that when performing aggregate functions over an array of Pair Types, the sum of those Pairs will be performed on the y property and the x will be ignored (and removed in the answer).

**Reply by josdejong** [reply]
It depends on how you recognize what a PairType is, your `isPairType` function. If this is ducktyped such that it checks whether dealing with an object containing `x`, `y`, and `isPairType=true` (or just `isPairType=true`) it should work I think, but if your check is something like `value instanceof PairType` it will return false when passing a plain object.

**Reply by cshaa** [reply]
Having a type that can represent either a point, or a horizontal line, or a vertical line, but nothing else, sounds like a really weird choice to me. If we get around to adding more comfortable tools for geometry to mathjs, I would imagine it more like this:
```javascript
plane = EuclideanSpace(2)
ptA = plane.point(2, 3)
ptB = plane.point(4, 5)
line1 = plane.line(A, B)
line2 = plane.axis[1]
```
However, experimenting with the parser and type checker sounds like fun, so I'm looking forward to the results of your work :)

**Reply by bernarma** [reply]
> Having a type that can represent either a point, or a horizontal line, or a vertical line, but nothing else, sounds like a really weird choice to me. If we get around to adding more comfortable tools for geometry to mathjs, I would imagine it more like this:
> 
> ```js
> plane = EuclideanSpace(2)
> ptA = plane.point(2, 3)
> ptB = plane.point(4, 5)
> line1 = plane.line(A, B)
> line2 = plane.axis[1]
> ```
> 
> However, experimenting with the parser and type checker sounds like fun, so I'm looking forward to the results of your work :)

It'll become clearer the more I work with it. I needed something to represent a "series" of data which is basically a collection of "points". I am hoping it becomes clearer the more I delve into it... I am currently working on the UI at the moment so no progress has been made that I can share - apart from the changes I made to the expression parser to support the new fundamental type.

Single Point: a = (1,3)
Array of Points: b = [(10,20), (90,10)]

My next step is to start incorporating this with my desired output which is a "Dataframe" / "Series". I also haven't called it a Point. I have called it a "Tuple" and it supports an ordered list of "n" items. Once it is clearer I will share the work - I am definitely not finished and something will come of this as I need this capability and I have searched high and low for a solution (without resorting to another language - I have tried Python and C++ as well - I like the simplicity of the single language but performance may decide when I start feeding large datasets into).

In regards to how much data I am envisaging I will need to handle: around 500,000 data points over a year per channel. There are approximately 10 channels that I need to deal with concurrently. This will represent a year, there is the option to change this range to a smaller value or longer value. Time will tell but I'll keep you posted with my progress (after I sort out the UI which is not one of my strengths).

**Reply by bernarma** [reply]
> Hi Mark,
> 
> A `Pair` can be useful but is relatively limited. I think it's similar to an Array or Matrix with just two values?
> 
> If you I want to hold more information together (x, y, date, ...), using an Object may be a good solution (that makes it explicit what value has what meaning, instead of relying on the position of a value in a Pair, Tuple, or Array.
> 
> Would something like this work for you?
> 
> ```
> var1 = { x: 1, y: 2 }
> var2 = { x: null, y: 3 }
> var2 = { x: 3, y: null }
> ```

I've been working with the concept a little more now, even to the point of creating this new data type called Pair although I am not 100% certain that is the way I want to go.

Sticking to out of the box functionality, I am going to investigate the object approach in more detail. One area I am looking for some advice on is the handling of this object for built-in functions such as add, sum, mean, etc.

Following this article, https://mathjs.org/examples/advanced/custom_datatype.js.html, I would like to extend the functionality of some of the core functions such as add, divide, sum, mean, etc. The issue I am finding is when I attempt to create a new function when dealing with arrays I get the following error:

```
[Nest] 33056   - 06/29/2021, 4:26:07 PM   [ExceptionsHandler] Signature "Array" is defined twice +3130ms
Error: Signature "Array" is defined twice
    at validateUnique (/Users/berniemac/Develop/projekt1/mathjs/node_modules/typed-function/typed-function.js:1288:17)
    at extractSignatures (/Users/berniemac/Develop/projekt1/mathjs/node_modules/typed-function/typed-function.js:1303:15)
    at Function....Function (/Users/berniemac/Develop/projekt1/mathjs/node_modules/typed-function/typed-function.js:1336:50)
    at Function.preprocessRestParams (/Users/berniemac/Develop/projekt1/mathjs/node_modules/typed-function/typed-function.js:781:28)
    at Function.generic (/Users/berniemac/Develop/projekt1/mathjs/node_modules/typed-function/typed-function.js:1073:27)
    at Function.typed (/Users/berniemac/Develop/projekt1/mathjs/node_modules/typed-function/typed-function.js:1092:24)
    at lazyTyped (/Users/berniemac/Develop/projekt1/mathjs/lib/cjs/core/create.js:194:23)
    at resolver (/Users/berniemac/Develop/projekt1/mathjs/lib/cjs/core/function/import.js:303:16)
    at Object.get [as mean] (/Users/berniemac/Develop/projekt1/mathjs/lib/cjs/utils/object.js:262:18)
    at _deleteTransform (/Users/berniemac/Develop/projekt1/mathjs/lib/cjs/core/function/import.js:220:53)
    at _importFactory (/Users/berniemac/Develop/projekt1/mathjs/lib/cjs/core/function/import.js:319:9)
    at Object.mathImport [as import] (/Users/berniemac/Develop/projekt1/mathjs/lib/cjs/core/function/import.js:130:11)
    at Object.fn (/Users/berniemac/Develop/projekt1/backend/dist/src/views/commands/handlers/evaluator.js:141:20)
    at TestHandler.<anonymous> (/Users/berniemac/Develop/projekt1/backend/dist/src/views/commands/handlers/test.handler.js:28:40)
    at Generator.next (<anonymous>)
    at /Users/berniemac/Develop/projekt1/backend/dist/src/views/commands/handlers/test.handler.js:14:71
```

I understand what the error is indicated but I need to handle the Array of "my type" but I would like to retain the existing implementation. Or maybe I am not understanding how the typed-function implementation works and you can indeed provide multiple functions depending on the type of the object within the array.

Any suggestions?

**Reply by josdejong** [reply]
This error occurs when you - like the message says - have defined the same signature twice, like:

```js
const myFunction = typed('myFunction', {
  'Array': (arr) => { ... },
  'Array': (arr) => { ... }
})
```

Typed function simply can't know which one of the two to execute. If you still need the original signature, you will have to replace the old signature with a new one which handles both cases some way. Alternatively, you can maybe create a differing type than `Array` so they don't conflict.
